src/cls/otp/cls_otp.cc

```diff
@@ -0,0 +1,510 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+/** \file
+ *
+ * This is an OSD class that implements methods for management
+ * and use of otp (one time password).
+ *
+ */
+
+#include <errno.h>
+#include <map>
+#include <list>
+
+#include <boost/range/adaptor/reversed.hpp>
+
+#include <liboath/oath.h>
+
+#include "include/types.h"
+#include "include/utime.h"
+#include "objclass/objclass.h"
+
+#include "common/errno.h"
+#include "common/Clock.h"
+
+#include "cls/otp/cls_otp_ops.h"
+#include "cls/otp/cls_otp_types.h"
+
+
+using namespace rados::cls::otp;
+
+
+CLS_VER(1,0)
+CLS_NAME(otp)
+
+#define ATTEMPTS_PER_WINDOW 5
+
+static string otp_header_key = "header";
+static string otp_key_prefix = "otp/";
+
+struct otp_header {
+  set<string> ids;
+
+  otp_header() {}
+
+  void encode(bufferlist &bl) const {
+    ENCODE_START(1, 1, bl);
+    ::encode(ids, bl);
+    ENCODE_FINISH(bl);
+  }
+  void decode(bufferlist::iterator &bl) {
+    DECODE_START(1, bl);
+    ::decode(ids, bl);
+    DECODE_FINISH(bl);
+  }
+};
+WRITE_CLASS_ENCODER(otp_header)
+
+struct otp_instance {
+  otp_info_t otp;
+
+  list<otp_check_t> last_checks; 
+  uint64_t last_success{0}; /* otp counter/step of last successful check */
+
+  otp_instance() {}
+
+  void encode(bufferlist &bl) const {
+    ENCODE_START(1, 1, bl);
+    ::encode(otp, bl);
+    ::encode(last_checks, bl);
+    ::encode(last_success, bl);
+    ENCODE_FINISH(bl);
+  }
+  void decode(bufferlist::iterator &bl) {
+    DECODE_START(1, bl);
+    ::decode(otp, bl);
+    ::decode(last_checks, bl);
+    ::decode(last_success, bl);
+    DECODE_FINISH(bl);
+  }
+
+  void trim_expired(const ceph::real_time& now);
+  void check(const string& token, const string& val, bool *update);
+  bool verify(const ceph::real_time& timestamp, const string& val);
+
+  void find(const string& token, otp_check_t *result);
+};
+WRITE_CLASS_ENCODER(otp_instance)
+
+
+void otp_instance::trim_expired(const ceph::real_time& now)
+{
+  ceph::real_time window_start = now - make_timespan(otp.step_size);
```

Comment:
`make_timespan()` takes a double, so this does an extra conversion. `std::chrono::seconds(otp.step_size)` may be more concise here?